Essa/GUI/Reactivity.hpp

```diff
@@ -0,0 +1,193 @@
+    }
+
+private:
+    Observable m_observable;
+    Mapper m_mapper;
+};
+
+}
+
+template<std::copyable T> class ObservableBase {
+public:
+    using Type = T;
+
+    ObservableBase() = default;
+    ObservableBase(ObservableBase const&) = delete;
+    ObservableBase(ObservableBase&&) = delete;
+    virtual ~ObservableBase() = default;
+
+    using ObserverFunc = std::function<void(T const&)>;
+
+    void observe(ObserverFunc&& observer) {
+        assert(observer);
+        // Inform about the initial value
+        observer(get());
+        m_observers.push_back(std::move(observer));
+    }
+
+    virtual T get() const = 0;
+
+    template<class Mapper> auto map(Mapper mapper) { return Reactivity::Map<decltype(*this), Mapper> { *this, std::move(mapper) }; }
+
+protected:
+    void notify(T const& state) {
+        for (auto const& obs : m_observers) {
+            obs(state);
+        }
+    }
+
+private:
+    std::vector<ObserverFunc> m_observers;
+};
+
+// Observable with constant value
+template<std::copyable T> class Observable : public ObservableBase<T> {
+public:
+    Observable(T&& t)
+        : m_value(std::forward<T>(t)) { }
+
+    Observable& operator=(T const& t) {
+        m_value = t;
+        notify(t);
+    }
+
+    virtual T get() const { return m_value; }
+
+private:
+    T m_value;
+};
+
+// Observable that calculates its value on the fly
+template<std::copyable T> class CalculatedObservable : public ObservableBase<T> {
+public:
+    using Calculator = std::function<T()>;
+
+    CalculatedObservable(Calculator&& t)
+        : m_calculator(std::move(t)) {
+        assert(m_calculator);
+    }
+
+    virtual T get() const override { return m_calculator(); }
+    using ObservableBase<T>::notify;
+
+private:
+    Calculator m_calculator;
+};
+
+// Read-only wrapper for Observable. Expose this if you want to make
+// some field public but read only
+template<std::copyable T> class ReadOnlyObservable {
+public:
+    using Type = T;
+    using ObserverFunc = std::function<void(T const&)>;
+
+    ReadOnlyObservable(ObservableBase<T>& ob)
+        : m_observable(ob) { }
+
+    void observe(ObserverFunc&& observer) { m_observable.observe(std::move(observer)); }
+    T const& get() const { return m_observable.m_value; }
+
+    template<class Mapper> auto map(Mapper mapper) {
+        return Reactivity::Map<std::remove_reference_t<decltype(*this)>, Mapper> {
+            *this,
+            std::move(mapper),
+        };
+    }
+
+private:
+    ObservableBase<T>& m_observable;
+};
+
+// Value that is read from somewhere. It may contain a function that
+// updates its value on every read.
+template<std::copyable T> class Read {
+public:
+    using Reader = std::function<T()>;
+
+    Read() = default;
+
+    // Construct from a function that will compute the value
+    // for every read.
+    Read& operator=(Reader&& t) {
+        assert(t);
+        // TODO: Deregister observer?
+        assert(!m_observing);
```

Comment:
`assert(t);` ensures that you don't pass a null std::function which is unfortunately possible in C++

`assert(!m_observing)` is because otherwise Read would be observing an Observable and a function at the same, leading to unpredictable behavior (in most cases taking value from Function, but sometimes from Observer when it runs `notify`)